src/shogun/mathematics/linalg/ratapprox/logdet/LogDetEstimator.cpp

```diff
@@ -145,71 +123,40 @@ SGVector<float64_t> CLogDetEstimator::sample(index_t num_estimates)
 		m_operator_log->get_operator()->get_dimension(),
 		m_trace_sampler->get_dimension());
 
-	// for storing the aggregators that submit_jobs return
-	CDynamicObjectArray* aggregators=new CDynamicObjectArray();
 	index_t num_trace_samples=m_trace_sampler->get_num_samples();
-
-	for (index_t i=0; i<num_estimates; ++i)
-	{
-		for (index_t j=0; j<num_trace_samples; ++j)
+	index_t idx_row = 0;
+	index_t idx_col = 0;
+	SGVector<float64_t> samples(num_estimates);
+	samples.zero();
+	// for omp
+#pragma omp parallel for
+		for (index_t i = 0; i < num_estimates; i++)
 		{
-			SG_INFO("Computing log-determinant trace sample %d/%d\n", j,
+			for (index_t j = 0; j < num_trace_samples; ++j)
+			{
+				SG_INFO(
+					"Computing log-determinant trace sample %d/%d\n", j,
 					num_trace_samples);
-
-			SG_DEBUG("Creating job for estimate %d, trace sample %d/%d\n", i, j,
-					num_trace_samples);
-			// get the trace sampler vector
-			SGVector<float64_t> s=m_trace_sampler->sample(j);
-			// create jobs with the sample vector and store the aggregator
-			CJobResultAggregator* agg=m_operator_log->submit_jobs(s);
-			aggregators->append_element(agg);
-			SG_UNREF(agg);
+				// get the trace sampler vector
+				SGVector<float64_t> s = m_trace_sampler->sample(j);
+				// calculate the result for sample s
+				float64_t result = m_operator_log->solve(s);
+#pragma omp critical // so that the dynamic array stays concurrent
+				{
+					samples[idx_col] += result;
```

Comment:
this could be reformulated instead of using critical using parallel for reduction...
check for example:
http://cs.umw.edu/~finlayson/class/fall16/cpsc425/notes/11-parfor.html

Reply:
Looks useful I will try to implement it.